### DISCUSSION TITLE
working WSL2 setup with docker-desktop

### DISCUSSION THREAD

**Discussion by sdktr** [srl-labs/containerlab#731] [Show and tell] [answered]
# Containerlab on WSL2 notes

Is someone able to verify my findings regarding containerlab on 'native' docker-desktop WSL2 setup? 

### Prep WSL2
WSL2 doesn't come with the right modules to run containerlabs network instructions. The [containerlab installation manual](https://containerlab.srlinux.dev/install/#fn:3) hints to a seperate 'real' docker-ce installation to overcome this.
I tried the workaround below, to add the modules required to my WSL2 kernel. A basic 2-node srl lab works with this setup.

>In case someone is looking for a way to load a module on WSL2: open /etc/modules-load.d/modules.conf and write the name of the kernel module you would like to load, one line per module. Then run wsl --shutdown, and wsl boot it up. Your module should be loaded. Tested on WSL2 Ubuntu focal.

Source: [Twitter discussion](https://github.com/kubernetes-sigs/kind/issues/707) links to [GH issue](https://github.com/kubernetes-sigs/kind/issues/707), that points to [StackExchange](https://unix.stackexchange.com/a/672846) which mentions [kind on WSL2 docs](https://kind.sigs.k8s.io/docs/user/using-wsl2/)

Name of the kernel module to add: ```xt_recent```

**Comment by hellt** [comment, marked as answer]
Thanks for this tip @sdktr 

I wonder why would you prefer to keep the docker desktop backend and not install docker natively on the WSL VM?

**Reply by bewing** [reply]
> I wonder why would you prefer to keep the docker desktop backend and not install docker natively on the WSL VM?

One reason to prefer Docker Desktop would be for a unified Docker environment between Powershell and WSL.  With Docker Desktop, I can access containers from either environment.  Additionally, I can use tools like kubectl in either environment to manage the same Docker Desktop k8s environment.  Finally, the network integration in Docker Desktop allows me to access k8s workloads from Chrome without messing with hosts files or similar, easing local development.

FWIW, with the latest MS WSL2 kernel (5.10.102.1, config file https://github.com/microsoft/WSL2-Linux-Kernel/blob/linux-msft-wsl-5.10.102.1/Microsoft/config-wsl), errors out on network create with
`Error: failed to lookup link "br-9060701c6c02": Link not found` when using Docker Desktop integration

**Reply by bewing** [reply]
If you are weird like me, you can also just run a second WSL2 distribution without Docker Desktop integration and docker-ce installed specifically for running labs.

**Comment by mulder999** [comment]
- The installer from [https://containerlab.dev/setup](https://containerlab.dev/setup) is not working on WSL2/debian (it does nothing nor print anything).
- Then I attempted to use docker-desktop integration and ended with the `Link not found` error. Probably because `docker0` interface does not exisist with Docker Desktop running in this mode. _The documentation let us think that a few additional steps could do but it is likely not the case..._
- Finally I installed docker-ce using the [https://get.docker.com](https://get.docker.com) script and ignored the warning `WSL DETECTED: We recommend using Docker Desktop for Windows.`

At this point, the example [srl02.clab.yml](https://github.com/srl-labs/containerlab/tree/main/lab-examples/srl02/srl02.clab.yml) fails with a permission issue `[FailedPrecondition] Setting permissions for file '/etc/opt/srlinux/config.tmp' failed with error_code = 1 [Operation not permitted]` but I assume it is an issue with the srlinux container which I do not plan to use anyway...

**Reply by prpa4901** [reply]
is this still an issue, I saw this issue even using docker-ce installed with my WSL2, right now if I point config.json as startup-script, didn't get this issue in the post-deploy action but then no json configs are getting copied to the sr_linux container. Next, I tried using cli file as startup-script option in the topology yaml, then I see this issue about permission issue for '/etc/opt/srlinux/config.tmp'. In both the case, the startup-script for copying the initial configs is not working. Let me know if I missing any step here. thanks

**Reply by hellt** [reply]
Try this https://github.com/srl-labs/wsl-containerlab

**Reply by prpa4901** [reply]
Thanks let me check this

**Reply by FloSch62** [reply]
I just made the latest clab release available in the 0.62.2-1.0 release, but you can also just do sudo containerlab version upgrade within older wsl clab releases
